Panic splitting pre-featurized datasets

On a pre-featurized dataset the following error is produced when creating models:

```
[ Jan 14 17:21:19 ] [ INFO ] checking folders `/home/chris/outputs/locust_medium_d3m-c482871493a0d0a4/train/datasetDoc.json` & `/home/chris/outputs/locust_medium_d3m-c482871493a0d0a4/test/datasetDoc.json` to see if the dataset has been previously split (api/compute/split.go:463)
panic: runtime error: index out of range [-1]

goroutine 25494027 [running]:
github.com/uncharted-distil/distil/api/compute.(*stratifiedSplitter).splitCategories(0xc0217d52c0, 0xffffffffffffffff, 0xc079544018, 0x439c, 0x439c, 0xc0012466b7)
        /home/chris/dev/git_work/distil/api/compute/split.go:254 +0x285
github.com/uncharted-distil/distil/api/compute.(*stratifiedSplitter).split(0xc0217d52c0, 0xc079544000, 0x439d, 0x439d, 0x439d, 0x439d, 0x0, 0x0, 0x0, 0x0, ...)
        /home/chris/dev/git_work/distil/api/compute/split.go:279 +0x315
github.com/uncharted-distil/distil/api/compute.SplitDataset(0xc004874190, 0x4a, 0x11f8b00, 0xc0217d52c0, 0x4a, 0xc000985401, 0xc0012466b7, 0x7, 0x3fe322d0e5604189, 0x0)
        /home/chris/dev/git_work/distil/api/compute/split.go:331 +0x475
github.com/uncharted-distil/distil/api/compute.(*SolutionRequest).PersistAndDispatch(0xc0010dbe60, 0xc0003101e0, 0x1219580, 0xc000eaa330, 0x1217c60, 0xc000eaa2d0, 0x1219ba0, 0xc000eaa2a0, 0x3fe999999999999a, 0x10001)
        /home/chris/dev/git_work/distil/api/compute/solution_request.go:758 +0xa0f
github.com/uncharted-distil/distil/api/ws.handleCreateSolutions(0xc000b74040, 0xc0003101e0, 0xc0003f8e70, 0xc0004189c0, 0xc0004189e0, 0xc00069af50)
        /home/chris/dev/git_work/distil/api/ws/pipeline.go:189 +0x56b
github.com/uncharted-distil/distil/api/ws.handleMessage(0xc000b74040, 0xc0003101e0, 0xc0003f8e70, 0xc0004189c0, 0xc0004189e0, 0xc0003f8ea0, 0xc00069af50)
        /home/chris/dev/git_work/distil/api/ws/pipeline.go:101 +0x285
created by github.com/uncharted-distil/distil/api/ws.handleSolutionMessage.func1
        /home/chris/dev/git_work/distil/api/ws/pipeline.go:92 +0x211
exit status 2
```

--- a/api/compute/search.go
+++ b/api/compute/search.go
@@ -290,8 +290,11 @@ func (s *SolutionRequest) dispatchSolutionSearchPipeline(statusChan chan Solutio
 
 		// persist results
 		log.Infof("persisting results in URI '%s'", resultURI)
-		s.persistSolutionResults(statusChan, client, solutionStorage, dataStorage, searchContext.searchID,
+		err = s.persistSolutionResults(statusChan, client, solutionStorage, dataStorage, searchContext.searchID,
 			searchContext.dataset, searchContext.storageName, searchSolutionID, fittedSolutionID, produceRequestID, resultID, resultURI)
+		if err != nil {
+			return nil, err
+		}
 	}
 	if err != nil {
 		return nil, err

--- a/api/compute/solution_request.go
+++ b/api/compute/solution_request.go
@@ -432,6 +432,8 @@ func createFitSolutionRequest(datasetURI string, fittedSolutionID string) *pipel
 }
 
 func (s *SolutionRequest) persistSolutionError(statusChan chan SolutionStatus, solutionStorage api.SolutionStorage, searchID string, solutionID string, err error) {
+	log.Errorf("solution '%s' errored: %v", solutionID, err)
+
 	// Check to see if this is a cancellation error and use a specific code for it if so
 	progress := compute.SolutionErroredStatus
 	cause := errors.Cause(err)
@@ -452,8 +454,6 @@ func (s *SolutionRequest) persistSolutionError(statusChan chan SolutionStatus, s
 		Error:      err,
 		Timestamp:  time.Now(),
 	}
-
-	log.Errorf("solution '%s' errored: %v", solutionID, err)
 }
 
 func (s *SolutionRequest) persistSolution(statusChan chan SolutionStatus, solutionStorage api.SolutionStorage, searchID string, solutionID string, explainedSolutionID string) {
@@ -517,28 +517,30 @@ func (s *SolutionRequest) persistRequestStatus(statusChan chan SolutionStatus, s
 
 func (s *SolutionRequest) persistSolutionResults(statusChan chan SolutionStatus, client *compute.Client, solutionStorage api.SolutionStorage,
 	dataStorage api.DataStorage, initialSearchID string, dataset string, storageName string, initialSearchSolutionID string,
-	fittedSolutionID string, produceRequestID string, resultID string, resultURI string) {
+	fittedSolutionID string, produceRequestID string, resultID string, resultURI string) error {
 	// persist the completed state
 	err := solutionStorage.PersistSolutionState(initialSearchSolutionID, compute.SolutionCompletedStatus, time.Now())
 	if err != nil {
 		// notify of error
 		s.persistSolutionError(statusChan, solutionStorage, initialSearchID, initialSearchSolutionID, err)
-		return
+		return err
 	}
 	// persist result metadata
 	err = solutionStorage.PersistSolutionResult(initialSearchSolutionID, fittedSolutionID, produceRequestID, "test", resultID, resultURI, compute.SolutionCompletedStatus, time.Now())
 	if err != nil {
 		// notify of error
 		s.persistSolutionError(statusChan, solutionStorage, initialSearchID, initialSearchSolutionID, err)
-		return
+		return err
 	}
 	// persist results
 	err = dataStorage.PersistResult(dataset, storageName, resultURI, s.TargetFeature.Key)
 	if err != nil {
 		// notify of error
 		s.persistSolutionError(statusChan, solutionStorage, initialSearchID, initialSearchSolutionID, err)
-		return
+		return err
 	}
+
+	return nil
 }
 
 func describeSolution(client *compute.Client, initialSearchSolutionID string) (*pipeline.DescribeSolutionResponse, error) {

--- a/api/model/dataset.go
+++ b/api/model/dataset.go
@@ -65,6 +65,7 @@ type Dataset struct {
 	LearningDataset string                 `json:"learningDataset"`
 	Clone           bool                   `json:"clone"`
 	Immutable       bool                   `json:"immutable"`
+	ParentDataset   string                 `json:"parentDataset"`
 }
 
 // QueriedDataset wraps dataset querying components into a single entity.

--- a/api/model/storage/elastic/dataset.go
+++ b/api/model/storage/elastic/dataset.go
@@ -51,6 +51,7 @@ func (s *Storage) CloneDataset(dataset string, datasetNew string, storageNameNew
 	ds.Name = datasetNew
 	ds.Folder = folderNew
 	ds.Source = metadata.Augmented
+	ds.ParentDataset = dataset
 	// cloned datasets CAN be altered
 	ds.Immutable = false
 	ds.Clone = true
@@ -77,6 +78,7 @@ func (s *Storage) UpdateDataset(dataset *api.Dataset) error {
 		"learningDataset": dataset.LearningDataset,
 		"clone":           dataset.Clone,
 		"immutable":       dataset.Immutable,
+		"parentDataset":   dataset.ParentDataset,
 	}
 
 	bytes, err := json.Marshal(source)
@@ -140,6 +142,7 @@ func (s *Storage) IngestDataset(datasetSource metadata.DatasetSource, meta *mode
 		"learningDataset":  meta.LearningDataset,
 		"clone":            meta.Clone,
 		"immutable":        meta.Immutable,
+		"parentDataset":    "",
 	}
 
 	bytes, err := json.Marshal(source)
@@ -209,6 +212,11 @@ func (s *Storage) parseDatasets(res *elastic.SearchResult, includeIndex bool, in
 		learningDataset, ok := json.String(src, "learningDataset")
 		if !ok {
 			learningDataset = ""
+		}
+		// extract the learning dataset
+		parentDataset, ok := json.String(src, "parentDataset")
+		if !ok {
+			parentDataset = ""
 		}
 		// extract the machine learned summary
 		summaryMachine, ok := json.String(src, "summaryMachine")
@@ -304,6 +312,7 @@ func (s *Storage) parseDatasets(res *elastic.SearchResult, includeIndex bool, in
 			LearningDataset: learningDataset,
 			Immutable:       immutable,
 			Clone:           clone,
+			ParentDataset:   parentDataset,
 		})
 	}
 	return datasets, nil

--- a/api/model/storage/elastic/storage.go
+++ b/api/model/storage/elastic/storage.go
@@ -207,6 +207,9 @@ func (s *Storage) InitializeMetadataStorage(overwrite bool) error {
 				"learningDataset": {
 					"type": "text"
 				},
+				"parentDataset": {
+					"type": "text"
+				},
 				"clone": {
 					"type": "boolean"
 				},

--- a/api/model/storage/postgres/result.go
+++ b/api/model/storage/postgres/result.go
@@ -272,7 +272,7 @@ func (s *Storage) PersistResult(dataset string, storageName string, resultURI st
 	}
 
 	// Translate from display name to storage name.
-	targetDisplayName, err := s.getDisplayName(dataset, targetVariable.Key)
+	targetHeaderName, err := s.getHeaderName(dataset, targetVariable.Key)
 	if err != nil {
 		return errors.Wrap(err, "unable to map target name")
 	}
@@ -283,9 +283,9 @@ func (s *Storage) PersistResult(dataset string, storageName string, resultURI st
 	for i, v := range records[0] {
 		fieldsHeader[v] = i
 	}
-	targetIndex, ok := fieldsHeader[targetDisplayName]
+	targetIndex, ok := fieldsHeader[targetHeaderName]
 	if !ok {
-		return errors.Wrapf(err, "unable to find target col '%s' in result header", targetDisplayName)
+		return errors.Wrapf(err, "unable to find target col '%s' in result header", targetHeaderName)
 	}
 	d3mIndexIndex, ok := fieldsHeader[model.D3MIndexFieldName]
 	if !ok {
@@ -980,20 +980,13 @@ func (s *Storage) getIsWeighted(resultURI string, weightTableName string) (bool,
 	return bool(values[0].(bool)), nil
 }
 
-func (s *Storage) getDisplayName(dataset string, key string) (string, error) {
-	displayName := ""
-	variables, err := s.metadata.FetchVariables(dataset, false, false, false)
+func (s *Storage) getHeaderName(dataset string, key string) (string, error) {
+	variable, err := s.metadata.FetchVariable(dataset, key)
 	if err != nil {
-		return "", errors.Wrap(err, "unable fetch variables for name mapping")
+		return "", errors.Wrap(err, "unable fetch variable for name mapping")
 	}
 
-	for _, v := range variables {
-		if v.Key == key {
-			displayName = v.DisplayName
-		}
-	}
-
-	return displayName, nil
+	return variable.HeaderName, nil
 }
 
 func mapFields(fields []*model.Variable) map[string]*model.Variable {

--- a/api/routes/save_dataset.go
+++ b/api/routes/save_dataset.go
@@ -89,7 +89,13 @@ func SaveDatasetHandler(metaCtor api.MetadataStorageCtor, dataCtor api.DataStora
 			handleError(w, err)
 			return
 		}
-		// version dataset
+
+		// update properties on the dataset (pull latest from store to pick up any other changes)
+		ds, err = metaStorage.FetchDataset(dataset, true, true, true)
+		if err != nil {
+			handleError(w, err)
+			return
+		}
 		ds.Immutable = true
 		// is no longer a clone due to the dropping of the filterParams
 		ds.Clone = false

--- a/api/routes/save_dataset.go
+++ b/api/routes/save_dataset.go
@@ -19,6 +19,7 @@ import (
 	"net/http"
 
 	"github.com/pkg/errors"
+	"github.com/uncharted-distil/distil-compute/model"
 	"github.com/uncharted-distil/distil/api/env"
 	api "github.com/uncharted-distil/distil/api/model"
 	"github.com/uncharted-distil/distil/api/task"
@@ -78,11 +79,21 @@ func SaveDatasetHandler(metaCtor api.MetadataStorageCtor, dataCtor api.DataStora
 			handleError(w, errors.New("can not mutate an immutable dataset"))
 			return
 		}
-		_, _, err = task.ExportDataset(dataset, metaStorage, dataStorage, invert, expandedFilterParams)
+
+		// export needs to invert the filters
+		filterModeReverse := map[string]string{model.IncludeFilter: model.ExcludeFilter, model.ExcludeFilter: model.IncludeFilter}
+		if expandedFilterParams != nil && expandedFilterParams.Highlight != nil {
+			expandedFilterParams.Highlight.Mode = filterModeReverse[expandedFilterParams.Highlight.Mode]
+		}
+		_, _, err = task.ExportDataset(dataset, metaStorage, dataStorage, !invert, expandedFilterParams)
 		if err != nil {
 			handleError(w, err)
 			return
 		}
+		if expandedFilterParams != nil && expandedFilterParams.Highlight != nil {
+			expandedFilterParams.Highlight.Mode = filterModeReverse[expandedFilterParams.Highlight.Mode]
+		}
+
 		// delete rows based on filterParams
 		err = dataStorage.SaveDataset(dataset, ds.StorageName, invert, expandedFilterParams)
 		if err != nil {

--- a/api/serialization/csv.go
+++ b/api/serialization/csv.go
@@ -45,18 +45,20 @@ func NewCSV() *CSV {
 
 // ReadDataset reads a raw dataset from the file system, loading the csv
 // data into memory.
-func (d *CSV) ReadDataset(uri string) (*api.RawDataset, error) {
-	data, err := d.ReadData(uri)
+func (d *CSV) ReadDataset(schemaFile string) (*api.RawDataset, error) {
+	meta, err := d.ReadMetadata(schemaFile)
 	if err != nil {
 		return nil, err
 	}
 
-	meta, err := d.ReadMetadata(uri)
+	data, err := d.ReadData(model.GetResourcePath(schemaFile, meta.GetMainDataResource()))
 	if err != nil {
 		return nil, err
 	}
 
 	return &api.RawDataset{
+		ID:       meta.ID,
+		Name:     meta.Name,
 		Data:     data,
 		Metadata: meta,
 	}, nil

--- a/api/serialization/parquet.go
+++ b/api/serialization/parquet.go
@@ -49,18 +49,20 @@ func NewParquet() *Parquet {
 
 // ReadDataset reads a raw dataset from the file system, loading the parquet
 // data into memory.
-func (d *Parquet) ReadDataset(uri string) (*api.RawDataset, error) {
-	data, err := d.ReadData(uri)
+func (d *Parquet) ReadDataset(schemaFile string) (*api.RawDataset, error) {
+	meta, err := d.ReadMetadata(schemaFile)
 	if err != nil {
 		return nil, err
 	}
 
-	meta, err := d.ReadMetadata(uri)
+	data, err := d.ReadData(model.GetResourcePath(schemaFile, meta.GetMainDataResource()))
 	if err != nil {
 		return nil, err
 	}
 
 	return &api.RawDataset{
+		ID:       meta.ID,
+		Name:     meta.Name,
 		Data:     data,
 		Metadata: meta,
 	}, nil

--- a/api/serialization/storage.go
+++ b/api/serialization/storage.go
@@ -78,18 +78,7 @@ func ReadDataset(schemaPath string) (*api.RawDataset, error) {
 	}
 
 	dataPath := model.GetResourcePath(schemaPath, meta.GetMainDataResource())
-	storage := GetStorage(dataPath)
-	data, err := storage.ReadData(dataPath)
-	if err != nil {
-		return nil, err
-	}
-
-	return &api.RawDataset{
-		Name:     meta.Name,
-		ID:       meta.ID,
-		Data:     data,
-		Metadata: meta,
-	}, nil
+	return GetStorage(dataPath).ReadDataset(schemaPath)
 }
 
 // WriteDataset determines which storage engine to use and then writes out the

--- a/api/serialization/storage.go
+++ b/api/serialization/storage.go
@@ -69,8 +69,8 @@ func GetParquetStorage() Storage {
 	return parquetStorage
 }
 
-// ReadData reads the metadata to find the main data reference, then reads that.
-func ReadData(schemaPath string) ([][]string, error) {
+// ReadDataset reads the metadata to find the main data reference, then reads that.
+func ReadDataset(schemaPath string) (*api.RawDataset, error) {
 	// metadata can be read by CSV storage
 	meta, err := csvStorage.ReadMetadata(schemaPath)
 	if err != nil {
@@ -84,5 +84,19 @@ func ReadData(schemaPath string) ([][]string, error) {
 		return nil, err
 	}
 
-	return data, nil
+	return &api.RawDataset{
+		Name:     meta.Name,
+		ID:       meta.ID,
+		Data:     data,
+		Metadata: meta,
+	}, nil
+}
+
+// WriteDataset determines which storage engine to use and then writes out the
+// metadata and the data using it.
+func WriteDataset(folderPath string, dataset *api.RawDataset) error {
+	// use the main data resource to determine the storage engine
+	storage := GetStorage(dataset.Metadata.GetMainDataResource().ResPath)
+
+	return storage.WriteDataset(folderPath, dataset)
 }

--- a/api/serialization/storage.go
+++ b/api/serialization/storage.go
@@ -68,3 +68,21 @@ func GetCSVStorage() Storage {
 func GetParquetStorage() Storage {
 	return parquetStorage
 }
+
+// ReadData reads the metadata to find the main data reference, then reads that.
+func ReadData(schemaPath string) ([][]string, error) {
+	// metadata can be read by CSV storage
+	meta, err := csvStorage.ReadMetadata(schemaPath)
+	if err != nil {
+		return nil, err
+	}
+
+	dataPath := model.GetResourcePath(schemaPath, meta.GetMainDataResource())
+	storage := GetStorage(dataPath)
+	data, err := storage.ReadData(dataPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}

--- a/api/task/dataset.go
+++ b/api/task/dataset.go
@@ -162,70 +162,83 @@ func updateLearningDataset(newDataset *api.RawDataset, metaDataset *api.Dataset,
 		return err
 	}
 
-	if parentDS.LearningDataset != "" {
-		// determine if there are new columns that were not part of the original dataset
-		parentVarMap := mapVariables(parentDS.Variables, func(variable *model.Variable) string { return variable.Key })
-		newVars := []*model.Variable{}
-		for _, v := range metaDataset.Variables {
-			if v.DistilRole == model.VarDistilRoleData {
-				if parentVarMap[v.Key] == nil {
-					newVars = append(newVars, v)
-				}
+	if parentDS.LearningDataset == "" {
+		return nil
+	}
+
+	// determine if there are new columns that were not part of the original dataset
+	parentVarMap := mapVariables(parentDS.Variables, func(variable *model.Variable) string { return variable.Key })
+	newVars := []*model.Variable{}
+	for _, v := range metaDataset.Variables {
+		if v.DistilRole == model.VarDistilRoleData {
+			if parentVarMap[v.Key] == nil {
+				newVars = append(newVars, v)
 			}
 		}
+	}
 
-		// copy the learning dataset
-		learningFolder := fmt.Sprintf("%s-featurized", newDataset.Metadata.ID)
-		learningFolder = path.Join(path.Dir(parentDS.LearningDataset), learningFolder)
-		err = util.Copy(parentDS.LearningDataset, learningFolder)
-		if err != nil {
-			return err
-		}
+	// copy the learning dataset
+	learningFolder := fmt.Sprintf("%s-featurized", newDataset.Metadata.ID)
+	learningFolder = path.Join(path.Dir(parentDS.LearningDataset), learningFolder)
+	err = util.Copy(parentDS.LearningDataset, learningFolder)
+	if err != nil {
+		return err
+	}
 
-		// read the prefeaturized data (need to load the metadata to read the data)
-		preFeaturizedDataset, err := serialization.ReadDataset(path.Join(learningFolder, compute.D3MDataSchema))
-		if err != nil {
-			return err
-		}
-		preFeaturizedMainDR := preFeaturizedDataset.Metadata.GetMainDataResource()
-		preFeaturizedVarMap := mapVariables(preFeaturizedMainDR.Variables, func(variable *model.Variable) string { return variable.Key })
-		preFeaturizedD3MIndex := preFeaturizedVarMap[model.D3MIndexFieldName].Index
+	// read the prefeaturized data (need to load the metadata to read the data)
+	preFeaturizedDataset, err := serialization.ReadDataset(path.Join(learningFolder, compute.D3MDataSchema))
+	if err != nil {
+		return err
+	}
+	preFeaturizedMainDR := preFeaturizedDataset.Metadata.GetMainDataResource()
+	preFeaturizedVarMap := mapVariables(preFeaturizedMainDR.Variables, func(variable *model.Variable) string { return variable.Key })
+	preFeaturizedD3MIndex := preFeaturizedVarMap[model.D3MIndexFieldName].Index
 
-		// add the missing columns row by row and only retain rows in the new dataset
-		// first build up the new variables by d3m index map
-		// then cycle through the featurized rows and append the variables
-		newDSD3MIndex := exportVarMap[model.D3MIndexFieldName].Index
-		newDataMap := map[string][]string{}
-		for _, r := range newDataset.Data[1:] {
-			newVarsData := []string{}
-			for i := 0; i < len(newVars); i++ {
-				newVarsData = append(newVarsData, r[newVars[i].Index])
-			}
-			newDataMap[r[newDSD3MIndex]] = newVarsData
-		}
-
-		// add the new fields to the metadata to generate the proper header
+	// add the missing columns row by row and only retain rows in the new dataset
+	// first build up the new variables by d3m index map
+	// then cycle through the featurized rows and append the variables
+	newDSD3MIndex := exportVarMap[model.D3MIndexFieldName].Index
+	newDataMap := map[string][]string{}
+	for _, r := range newDataset.Data[1:] {
+		newVarsData := []string{}
 		for i := 0; i < len(newVars); i++ {
-			newVar := newVars[i]
-			newVar.Index = i + len(preFeaturizedMainDR.Variables)
-			preFeaturizedMainDR.Variables = append(preFeaturizedMainDR.Variables, newVar)
+			newVarsData = append(newVarsData, r[newVars[i].Index])
 		}
+		newDataMap[r[newDSD3MIndex]] = newVarsData
+	}
 
-		preFeaturizedOutput := [][]string{preFeaturizedMainDR.GenerateHeader()}
-		for _, row := range preFeaturizedDataset.Data[1:] {
-			d3mIndexPre := row[preFeaturizedD3MIndex]
-			if newDataMap[d3mIndexPre] != nil {
-				rowComplete := append(row, newDataMap[d3mIndexPre]...)
-				preFeaturizedOutput = append(preFeaturizedOutput, rowComplete)
-			}
-		}
+	// add the new fields to the metadata to generate the proper header
+	for i := 0; i < len(newVars); i++ {
+		newVar := newVars[i]
+		newVar.Index = i + len(preFeaturizedMainDR.Variables)
+		preFeaturizedMainDR.Variables = append(preFeaturizedMainDR.Variables, newVar)
+	}
 
-		// output the new pre featurized data
-		preFeaturizedDataset.Data = preFeaturizedOutput
-		err = serialization.WriteDataset(learningFolder, preFeaturizedDataset)
-		if err != nil {
-			return err
+	preFeaturizedOutput := [][]string{preFeaturizedMainDR.GenerateHeader()}
+	for _, row := range preFeaturizedDataset.Data[1:] {
+		d3mIndexPre := row[preFeaturizedD3MIndex]
+		if newDataMap[d3mIndexPre] != nil {
+			rowComplete := append(row, newDataMap[d3mIndexPre]...)
+			preFeaturizedOutput = append(preFeaturizedOutput, rowComplete)
 		}
+	}
+
+	// output the new pre featurized data
+	preFeaturizedDataset.Data = preFeaturizedOutput
+	err = serialization.WriteDataset(learningFolder, preFeaturizedDataset)
+	if err != nil {
+		return err
+	}
+
+	// update the learning dataset for the new dataset
+	metaDataset, err = metaStorage.FetchDataset(metaDataset.ID, true, true, true)
+	if err != nil {
+		return err
+	}
+	metaDataset.LearningDataset = learningFolder
+	err = metaStorage.UpdateDataset(metaDataset)
+	if err != nil {
+		return err
 	}
 
 	return nil

--- a/api/task/dataset.go
+++ b/api/task/dataset.go
@@ -204,20 +204,20 @@ func updateLearningDataset(newDataset *api.RawDataset, metaDataset *api.Dataset,
 			newDataMap[r[newDSD3MIndex]] = newVarsData
 		}
 
-		preFeaturizedOutput := [][]string{}
+		// add the new fields to the metadata to generate the proper header
+		for i := 0; i < len(newVars); i++ {
+			newVar := newVars[i]
+			newVar.Index = i + len(preFeaturizedMainDR.Variables)
+			preFeaturizedMainDR.Variables = append(preFeaturizedMainDR.Variables, newVar)
+		}
+
+		preFeaturizedOutput := [][]string{preFeaturizedMainDR.GenerateHeader()}
 		for _, row := range preFeaturizedDataset.Data[1:] {
 			d3mIndexPre := row[preFeaturizedD3MIndex]
 			if newDataMap[d3mIndexPre] != nil {
 				rowComplete := append(row, newDataMap[d3mIndexPre]...)
 				preFeaturizedOutput = append(preFeaturizedOutput, rowComplete)
 			}
-		}
-
-		// add the new fields to the metadata
-		for i := 0; i < len(newVars); i++ {
-			newVar := newVars[i]
-			newVar.Index = i + len(preFeaturizedMainDR.Variables)
-			preFeaturizedMainDR.Variables = append(preFeaturizedMainDR.Variables, newVar)
 		}
 
 		// output the new pre featurized data

--- a/api/task/dataset.go
+++ b/api/task/dataset.go
@@ -185,7 +185,7 @@ func updateLearningDataset(newDataset *api.RawDataset, metaDataset *api.Dataset,
 		return err
 	}
 
-	// read the prefeaturized data (need to load the metadata to read the data)
+	// read the prefeaturized data
 	preFeaturizedDataset, err := serialization.ReadDataset(path.Join(learningFolder, compute.D3MDataSchema))
 	if err != nil {
 		return err
@@ -222,6 +222,11 @@ func updateLearningDataset(newDataset *api.RawDataset, metaDataset *api.Dataset,
 			preFeaturizedOutput = append(preFeaturizedOutput, rowComplete)
 		}
 	}
+
+	// prefeaturized metadata needs to match the new dataset
+	preFeaturizedDataset.Metadata.ID = newDataset.Metadata.ID
+	preFeaturizedDataset.Metadata.Name = newDataset.Metadata.Name
+	preFeaturizedDataset.Metadata.StorageName = newDataset.Metadata.StorageName
 
 	// output the new pre featurized data
 	preFeaturizedDataset.Data = preFeaturizedOutput

--- a/api/task/dataset.go
+++ b/api/task/dataset.go
@@ -119,6 +119,9 @@ func ExportDataset(dataset string, metaStorage api.MetadataStorage, dataStorage 
 	if err != nil {
 		return "", "", err
 	}
+
+	// need to update metadata variable order to match extracted data
+
 	// update the header with the proper variable names
 	data[0] = meta.GetMainDataResource().GenerateHeader()
 	dataRaw := &api.RawDataset{
@@ -134,6 +137,64 @@ func ExportDataset(dataset string, metaStorage api.MetadataStorage, dataStorage 
 	err = storage.WriteDataset(outputFolder, dataRaw)
 	if err != nil {
 		return "", "", err
+	}
+
+	// need to write the prefeaturized version of the dataset if it exists
+	if metaDataset.ParentDataset != "" {
+		parentDS, err := metaStorage.FetchDataset(metaDataset.ParentDataset, false, false, false)
+		if err != nil {
+			return "", "", err
+		}
+
+		if parentDS.LearningDataset != "" {
+			// determine if there are new columns that were not part of the original dataset
+
+			parentVarMap := mapVariables(parentDS.Variables, func(variable *model.Variable) string { return variable.Key })
+			newVars := []*model.Variable{}
+			for _, v := range metaDataset.Variables {
+				if v.DistilRole == model.VarDistilRoleData {
+					if parentVarMap[v.Key] == nil {
+						newVars = append(newVars, v)
+					}
+				}
+			}
+
+			// copy the learning dataset
+			learningFolder := fmt.Sprintf("%s-featurized", meta.ID)
+			learningFolder = path.Join(path.Dir(parentDS.LearningDataset), learningFolder)
+			err = util.Copy(parentDS.LearningDataset, learningFolder)
+			if err != nil {
+				return "", "", err
+			}
+
+			// read the prefeaturized data (need to load the metadata to read the data)
+			preFeaturizedData, err := serialization.ReadData(path.Join(learningFolder, compute.D3MDataSchema))
+			if err != nil {
+				return "", "", err
+			}
+
+			// add the missing columns row by row
+			newDSD3MIndex := -1
+			for i, v := range data[0] {
+				if v == model.D3MIndexFieldName {
+					newDSD3MIndex = i
+				}
+			}
+			parentD3MIndex := parentDS.GetD3MIndexVariable()
+
+			newDataMap := map[string][]string{}
+			for _, r := range data[1:] {
+				newVarsData := []string{}
+				for i := 0; i < len(newVars); i++ {
+					newVarsData = append(newVarsData, r[newVars[i].Index])
+				}
+				newDataMap[r[newDSD3MIndex]] = newVarsData
+			}
+			for i, row := range preFeaturizedData[1:] {
+				log.Infof("something %v%v%v", i, row, parentD3MIndex)
+			}
+
+		}
 	}
 
 	return dataset, outputFolder, err

--- a/api/task/dataset.go
+++ b/api/task/dataset.go
@@ -121,6 +121,15 @@ func ExportDataset(dataset string, metaStorage api.MetadataStorage, dataStorage 
 	}
 
 	// need to update metadata variable order to match extracted data
+	header := data[0]
+	exportedVariables := make([]*model.Variable, len(header))
+	exportVarMap := mapVariables(metaDataset.Variables, func(variable *model.Variable) string { return variable.Key })
+	for i, v := range header {
+		variable := exportVarMap[v]
+		variable.Index = i
+		exportedVariables[i] = variable
+	}
+	meta.GetMainDataResource().Variables = exportedVariables
 
 	// update the header with the proper variable names
 	data[0] = meta.GetMainDataResource().GenerateHeader()
@@ -141,63 +150,85 @@ func ExportDataset(dataset string, metaStorage api.MetadataStorage, dataStorage 
 
 	// need to write the prefeaturized version of the dataset if it exists
 	if metaDataset.ParentDataset != "" {
-		parentDS, err := metaStorage.FetchDataset(metaDataset.ParentDataset, false, false, false)
-		if err != nil {
-			return "", "", err
-		}
-
-		if parentDS.LearningDataset != "" {
-			// determine if there are new columns that were not part of the original dataset
-
-			parentVarMap := mapVariables(parentDS.Variables, func(variable *model.Variable) string { return variable.Key })
-			newVars := []*model.Variable{}
-			for _, v := range metaDataset.Variables {
-				if v.DistilRole == model.VarDistilRoleData {
-					if parentVarMap[v.Key] == nil {
-						newVars = append(newVars, v)
-					}
-				}
-			}
-
-			// copy the learning dataset
-			learningFolder := fmt.Sprintf("%s-featurized", meta.ID)
-			learningFolder = path.Join(path.Dir(parentDS.LearningDataset), learningFolder)
-			err = util.Copy(parentDS.LearningDataset, learningFolder)
-			if err != nil {
-				return "", "", err
-			}
-
-			// read the prefeaturized data (need to load the metadata to read the data)
-			preFeaturizedData, err := serialization.ReadData(path.Join(learningFolder, compute.D3MDataSchema))
-			if err != nil {
-				return "", "", err
-			}
-
-			// add the missing columns row by row
-			newDSD3MIndex := -1
-			for i, v := range data[0] {
-				if v == model.D3MIndexFieldName {
-					newDSD3MIndex = i
-				}
-			}
-			parentD3MIndex := parentDS.GetD3MIndexVariable()
-
-			newDataMap := map[string][]string{}
-			for _, r := range data[1:] {
-				newVarsData := []string{}
-				for i := 0; i < len(newVars); i++ {
-					newVarsData = append(newVarsData, r[newVars[i].Index])
-				}
-				newDataMap[r[newDSD3MIndex]] = newVarsData
-			}
-			for i, row := range preFeaturizedData[1:] {
-				log.Infof("something %v%v%v", i, row, parentD3MIndex)
-			}
-
-		}
+		err = updateLearningDataset(dataRaw, metaDataset, exportVarMap, metaStorage)
 	}
 
 	return dataset, outputFolder, err
+}
+
+func updateLearningDataset(newDataset *api.RawDataset, metaDataset *api.Dataset, exportVarMap map[string]*model.Variable, metaStorage api.MetadataStorage) error {
+	parentDS, err := metaStorage.FetchDataset(metaDataset.ParentDataset, false, false, false)
+	if err != nil {
+		return err
+	}
+
+	if parentDS.LearningDataset != "" {
+		// determine if there are new columns that were not part of the original dataset
+		parentVarMap := mapVariables(parentDS.Variables, func(variable *model.Variable) string { return variable.Key })
+		newVars := []*model.Variable{}
+		for _, v := range metaDataset.Variables {
+			if v.DistilRole == model.VarDistilRoleData {
+				if parentVarMap[v.Key] == nil {
+					newVars = append(newVars, v)
+				}
+			}
+		}
+
+		// copy the learning dataset
+		learningFolder := fmt.Sprintf("%s-featurized", newDataset.Metadata.ID)
+		learningFolder = path.Join(path.Dir(parentDS.LearningDataset), learningFolder)
+		err = util.Copy(parentDS.LearningDataset, learningFolder)
+		if err != nil {
+			return err
+		}
+
+		// read the prefeaturized data (need to load the metadata to read the data)
+		preFeaturizedDataset, err := serialization.ReadDataset(path.Join(learningFolder, compute.D3MDataSchema))
+		if err != nil {
+			return err
+		}
+		preFeaturizedMainDR := preFeaturizedDataset.Metadata.GetMainDataResource()
+		preFeaturizedVarMap := mapVariables(preFeaturizedMainDR.Variables, func(variable *model.Variable) string { return variable.Key })
+		preFeaturizedD3MIndex := preFeaturizedVarMap[model.D3MIndexFieldName].Index
+
+		// add the missing columns row by row and only retain rows in the new dataset
+		// first build up the new variables by d3m index map
+		// then cycle through the featurized rows and append the variables
+		newDSD3MIndex := exportVarMap[model.D3MIndexFieldName].Index
+		newDataMap := map[string][]string{}
+		for _, r := range newDataset.Data[1:] {
+			newVarsData := []string{}
+			for i := 0; i < len(newVars); i++ {
+				newVarsData = append(newVarsData, r[newVars[i].Index])
+			}
+			newDataMap[r[newDSD3MIndex]] = newVarsData
+		}
+
+		preFeaturizedOutput := [][]string{}
+		for _, row := range preFeaturizedDataset.Data[1:] {
+			d3mIndexPre := row[preFeaturizedD3MIndex]
+			if newDataMap[d3mIndexPre] != nil {
+				rowComplete := append(row, newDataMap[d3mIndexPre]...)
+				preFeaturizedOutput = append(preFeaturizedOutput, rowComplete)
+			}
+		}
+
+		// add the new fields to the metadata
+		for i := 0; i < len(newVars); i++ {
+			newVar := newVars[i]
+			newVar.Index = i + len(preFeaturizedMainDR.Variables)
+			preFeaturizedMainDR.Variables = append(preFeaturizedMainDR.Variables, newVar)
+		}
+
+		// output the new pre featurized data
+		preFeaturizedDataset.Data = preFeaturizedOutput
+		err = serialization.WriteDataset(learningFolder, preFeaturizedDataset)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 // UpdateExtremas will update every field's extremas in the specified dataset.

